CIP2/CIP2.md

```diff
@@ -0,0 +1,1001 @@
+
+### Coverage of Payments
+
+This property states that the total value of _inputs_ in the resulting [coin
+selection](#coin-selection) result is sufficient to _cover_ the total value of
+the [requested output set](#requested-output-set).
+
+In particular:
+
+  * **_v_**<sub>selected</sub> ≥ **_v_**<sub>requested</sub>
+
+Where:
+
+  * **_v_**<sub>requested</sub>
+
+    is the total value of the [requested output set](#requested-output-set)
+
+
+  * **_v_**<sub>selected</sub>
+
+    is the total value of the _inputs_ field of the [coin
+    selection](#coin-selection) result.
+
+### Correctness of Change
+
+This property states that the correct amount of _change_ was generated.
+
+In particular:
+
+  * **_v_**<sub>selected</sub>
+  = **_v_**<sub>requested</sub> + **_v_**<sub>change</sub>
+
+Where:
+
+  * **_v_**<sub>change</sub>
+
+    is the total value of the _change_ field of the [coin
+    selection](#coin-selection) result.
+
+  * **_v_**<sub>requested</sub>
+
+    is the total value of the [requested output set](#requested-output-set)
+
+  * **_v_**<sub>selected</sub>
+
+    is the total value of the _inputs_ field of the [coin
+    selection](#coin-selection) result.
+
+### Conservation of UTxO
+
+This property states that every entry in the [initial UTxO
+set](#initial-utxo-set) is included in _either_ the inputs set of the generated
+[coin selection](#coin-selection), _or_ in the [remaining UTxO
+set](#remaining-utxo-set), but _not both_.
+
+  * If a UTxO entry _is_ selected by the coin selection algorithm, it is
+    included in the [coin selection](#coin-selection) inputs set.
+
+  * If a UTxO entry is _not_ selected by the coin selection algorithm, it is
+    included in the [remaining UTxO set](#remaining-utxo-set).
+
+The following laws hold:
+
+  * **_U_**<sub>initial  </sub> ⊃ **_U_**<sub>remaining</sub>
+  * **_U_**<sub>initial  </sub> ⊇ **_U_**<sub>selected </sub>
+
+And:
+
+  * **_U_**<sub>remaining</sub> ∩ **_U_**<sub>selected </sub> = ∅
+  * **_U_**<sub>remaining</sub> ⋃ **_U_**<sub>selected </sub> =
+    **_U_**<sub>initial  </sub>
+
+Where:
+
+  * **_U_**<sub>initial</sub>
+
+    is the [initial UTxO set](#initial-utxo-set).
+
+  * **_U_**<sub>remaining</sub>
+
+    is the [remaining UTxO set](#remaining-utxo-set).
+
+  * **_U_**<sub>selected</sub>
+
+    is the value of the _inputs_ field of the [coin selection](#coin-selection)
+    result.
+
+### Conservation of Outputs
+
+This property states that the [requested output set](#requested-output-set)
+is _conserved_ in the [coin selection](#coin-selection) result.
+
+In particular, the _outputs_ field of the [coin selection](#coin-selection)
+result should be _equal to_ the [requested output set](#requested-output-set).
+
+## Failure Modes
+
+There are a number of ways in which a coin selection algorithm can fail:
+
+  * #### UTxO Balance Insufficient
+
+    This failure occurs when the total value of the entries within the [initial
```

Comment:
@SebastienGllmt wrote:
> It may be worth noting that this can also happen in the case that UTXO in your wallet contains less ADA than the cost of including it in the input. The sum of the dust may have a sufficient balance, but each input individually is not worth adding.

That is indeed true, and a very good point.

However, in Cardano Wallet, dust elimination is mainly the responsibility of the _fee adjustment_ algorithm, whereas this document describes the algorithms involved in making an _initial coin selection_.

See https://github.com/cardano-foundation/CIPs/pull/2#discussion_r464806172 for an explanation of these two phases (which are consecutive and distinct from one another).